packages/zowe-explorer/src/dataset/DatasetTree.ts

```diff
@@ -1513,9 +1515,15 @@ export class DatasetTree extends ZoweTreeProvider implements IZoweTree<IZoweData
      * @param this (resolves ESlint warning about unbound methods)
      * @param doc A doc URI that was closed
      */
-    public static onDidCloseTextDocument(this: void, doc: vscode.TextDocument): void {
+    public static onDidCloseTextDocument(this: void, doc: vscode.TextDocument): Promise<void> {
         if (doc.uri.fsPath.includes(globals.DS_DIR)) {
-            updateOpenFiles(TreeProviders.ds, doc.uri.fsPath, null);
+            return ZoweSaveQueue.all().then(() => {
+                if (!vscode.workspace.textDocuments.find(({ uri }) => uri.fsPath === doc.uri.fsPath)?.isDirty) {
+                    // Remove document from cache only if there are no pending unsaved changes
+                    updateOpenFiles(TreeProviders.ds, doc.uri.fsPath, null);
+                    LocalFileManagement.removeRecoveredFile(doc);
+                }
+            });
```

Comment:
I think we discussed this offline at one point - but what happens if one of the queued save operations fails? Should we still run this logic in a `.catch` block to see if the given URI was saved anyway?

Reply:
The ZoweSaveQueue should catch and handle any errors during save operations, so I don't think we need to catch them here.
https://github.com/zowe/zowe-explorer-vscode/blob/0ddb5d94f937cb151a54189c5ca2226f3d2ddf3d/packages/zowe-explorer/src/abstract/ZoweSaveQueue.ts#L62

Reply:
Thanks for the clarification!